### PR TITLE
promoting member to applicant

### PR DESCRIPTION
**Pull request by opqdonut**
for #2581

# Checklist for author

Remove items that aren't applicable, check items that are done.

## Reviewability
- [x] Link to issue
- [x] Consider adding screenshots for ease of review

## Backwards compatibility
- [x] API is backwards compatible or completely new
- [x] Events are backwards compatible _or_ have migrations

## Documentation
- [x] Update changelog if necessary
- [x] API is documented and shows up in Swagger UI
- [x] Components are added to guide page
- [x] Update manual/ (if applicable)

## Testing
- [x] Complex logic is unit tested
- [ ] Valuable features are integration / browser / acceptance tested automatically

### CODE REVIEW

File: src/clj/rems/application/master_workflow.clj
Comment (opqdonut) [inline src/clj/rems/application/master_workflow.clj:59]: changing the applicant of a returned application seems a bit weird, but perhaps the application has been returned to e.g. a user who can't log in any more?

**Comment by opqdonut** [comment]
make-applicant

-----

make-applicant-2

**Comment by opqdonut** [comment]
do you think there should be a browser test?

File: test/clj/rems/api/test_applications.clj
Comment (Macroz) [inline test/clj/rems/api/test_applications.clj:758]: Could introduce the `ok-command` variant of `send-command` that does the comparison?

File: resources/translations/en.edn
Comment (Macroz) [inline resources/translations/en.edn:40]: I don't like this mismatch here either i.e. promote vs. make or change. We should decide on one name and be consistent, hopefully up to the event level?

File: src/cljs/rems/application.cljs
Comment (Macroz) [inline src/cljs/rems/application.cljs:662]: This is starting to feel excessive, combined with the need to name and pass these forwards. Maybe we could have a higher order (container) component `with-permissions` that would add to props this kind of kw-parameters? Or use lower level (container) components that do roles checking?

File: CHANGELOG.md
Comment (Macroz) [inline CHANGELOG.md:33]: Hmm which should be the default?

**Review by Macroz — APPROVED** [review]
Looks pretty complete. I think I would like to discuss the points I added before we merge. Like the naming of the command especially. Other things found were really small.

File: test/clj/rems/application/test_model.clj
Comment (opqdonut) [reply test/clj/rems/application/test_model.clj:1087]: Yeah, it is. The command handler generates a `:user-not-member` error in this case. However as per usual, we shouldn't fail to apply an event.

Comment (Macroz) [reply test/clj/rems/application/test_model.clj:1087]: Yeah, noticed that much later in the commits. I think the reasoning, not to limit events, is good.

File: src/clj/rems/application/commands.clj
Comment (opqdonut) [reply src/clj/rems/application/commands.clj:77]: Yeah I wanted to name the command & event differently because I thought the event might get generated from other sources as well. Also the command is more limited in scope than the event. I'd be fine with `ChangeApplicant` or `MemberToApplicant`.

File: test/clj/rems/api/test_applications.clj
Comment (opqdonut) [reply test/clj/rems/api/test_applications.clj:758]: Hmm, good suggestion. One argument against is that with an `(is (= ` you get slightly nicer errors (like "expected `{:success true}` but was `{:success false :errors [{:type :forbidden}]}`)

File: src/clj/rems/application/commands.clj
Comment (Macroz) [reply src/clj/rems/application/commands.clj:77]: - Promote sounds to me that there can be several applicants after it. 
- Change applicant sounds like the old is not anymore (matches current implementation).
- Member to applicant could support multiple applicants as well.

Maybe change is the best match of these three ideas?

File: resources/translations/en.edn
Comment (opqdonut) [reply resources/translations/en.edn:40]: I thought promote to applicant is a clear way of expressing the command for devlopers / API users / technical folk, but users are fine with a slightly more fuzzy "Make applicant", especially since they only see the button in context.

File: src/cljs/rems/application.cljs
Comment (opqdonut) [reply src/cljs/rems/application.cljs:662]: I tried to do a refactoring where we'd pass `permissions` directly to `member-info`, but unfortunately we have some special cases like the applicant (can't be removed) or invited members (can't be promoted even if promote-to-applicant is in permissions).

I guess I could push the handling of these special cases inside `member-info` as it already has the `invited-user?` parameter. In that case we could just pass `permissions` down.

File: resources/translations/en.edn
Comment (Macroz) [reply resources/translations/en.edn:40]: Sounds good, and matches the idea of having only one applicant after.

File: CHANGELOG.md
Comment (opqdonut) [reply CHANGELOG.md:33]: config-defaults.edn doesn't disable any commands currently, so I thought I'd follow that pattern. One alternative would be to have an explicit feature flag for this feature if we consider it experimental.

File: src/clj/rems/application/commands.clj
Comment (Macroz) [reply src/clj/rems/application/commands.clj:77]: MakeApplicant?

File: resources/translations/en.edn
Comment (Macroz) [reply resources/translations/en.edn:40]: Actually on second thought it is ambivalent

File: src/clj/rems/application/master_workflow.clj
Comment (opqdonut) [reply src/clj/rems/application/master_workflow.clj]: I felt like changing the applicant of a returned application would have more corner cases to test, but perhaps it's not so.

File: src/cljs/rems/application.cljs
Comment (opqdonut) [reply src/cljs/rems/application.cljs:662]: Yeah, I'll try this refactor out and push it if it feels good.

Comment (opqdonut) [reply src/cljs/rems/application.cljs]: The refactor came out pretty nicely I think!

File: src/clj/rems/application/master_workflow.clj
Comment (Macroz) [reply src/clj/rems/application/master_workflow.clj]: At least, there is no need for a `past-applicants`?

File: test/clj/rems/api/test_applications.clj
Comment (Macroz) [reply test/clj/rems/api/test_applications.clj:758]: Hmm how do we manage in the other places with this?

File: src/cljs/rems/application.cljs
Comment (opqdonut) [inline src/cljs/rems/application.cljs:622]: Because member-info gets the whole application, I was able to just push all the logic inside it. It might be better to split this into a view component and a container that does the logic, but then we'd be backin the world of `can-remove?` parameters.

File: CHANGELOG.md
Comment (Macroz) [reply CHANGELOG.md:33]: We should let @jaakkocsc decide the default (is it ok to add as default new command, and perhaps disable by hand in the environments CSC manages)

File: test/clj/rems/api/test_applications.clj
Comment (opqdonut) [reply test/clj/rems/api/test_applications.clj:758]: This file uses the `(is (= {:success true} ...)` a lot

File: src/cljs/rems/application.cljs
Comment (Macroz) [reply src/cljs/rems/application.cljs:622]: I think there are some higher order components that would make sense in the future, but no need to complicate this PR. I'll add an idea task.

File: src/clj/rems/application/master_workflow.clj
Comment (opqdonut) [reply src/clj/rems/application/master_workflow.clj]: Yeah, I don't think there is a need for past-applicants currently, and it's easy to add to the projection later if needed.

File: src/cljs/rems/application.cljs
Comment (Macroz) [reply src/cljs/rems/application.cljs:622]: https://github.com/CSCfi/rems/issues/2659

File: test/clj/rems/api/test_applications.clj
Comment (Macroz) [reply test/clj/rems/api/test_applications.clj:758]: Potentially there could be some machinery for assertions or exception messages here like `assert-response-is-ok` or so but I don't think it makes this much better. Usually there are other fields to check so an `is map` is fine. It seems to be this file that does this a lot but I guess not enough to matter.

Comment (Macroz) [reply test/clj/rems/api/test_applications.clj:758]: Also not worth a special macro :)

File: src/clj/rems/application/commands.clj
Comment (opqdonut) [reply src/clj/rems/application/commands.clj:77]: Renamed to ChangeApplicant!

File: CHANGELOG.md
Comment (opqdonut) [reply CHANGELOG.md:33]: Discussed this on chat, agreed with Jaakko that enabling by default is ok. We'll just point people at the changelog and let them test the new feature in their demo environment before deciding whether to disable it or not. Enabling by default is simpler codewise.

**Pull request merged**
